### PR TITLE
Remote Port Forward - 2.2 Updates

### PR DESCRIPTION
**Pull request by thiagomayllart**
Hello everyone! I've recently updated the port forward module to fit the 2.2 updates. It is working the same way as the previous pull request in: #18

### CODE REVIEW

File: Payload_Type/Apollo/agent_code/Apollo/C2Profiles/DefaultProfile.cs [hunk 1]
```diff
@@ -473,13 +487,19 @@ override public Mythic.Structs.TaskQueue GetMessages(Apollo.Agent agent)
                 {
                     response.SocksDatagrams = resp.socks;
                 }
+                if (resp.rportfwds != null)
+                {
+                    response.PortFwdDg = new PortFwdDatagram[1];
+                    response.PortFwdDg[0].data = JsonConvert.DeserializeObject<Dictionary<String, Dictionary<String, Dictionary<String, Dictionary<String, Dictionary<int,String>>>>>>(resp.rportfwds[0]);
```
Comment (djhohnstein) [inline Payload_Type/Apollo/agent_code/Apollo/C2Profiles/DefaultProfile.cs:493]: This datagram should be changed to a struct for code maintainability/readability. Five nested dictionaries, especially in subsequent function calls, starts to make it difficult to know how things are going wrong.

File: Payload_Type/Apollo/agent_code/Apollo/Mythic.cs [hunk 2]
```diff
@@ -301,6 +306,16 @@ public struct SocksDatagram
             public string data;
         }
 
+        public struct PortFwdDatagram
+        {
+            public Dictionary<String,Dictionary<String,Dictionary<String,Dictionary<String,Dictionary<int,String>>>>> data;
```
Comment (djhohnstein) [inline Payload_Type/Apollo/agent_code/Apollo/Mythic.cs:311]: Same comment as earlier - this should probably be some form of nested structure so this is easier to grok

File: Payload_Type/Apollo/agent_code/Apollo/RPortFwdProxy/Classes/ProxyConnection.cs [hunk 3]
```diff
@@ -0,0 +1,276 @@
...
+				}
+                            }
+			}
+                    }
+                }
+	    // keep reading from operatorMapQueue and send to operatorMapConn
+            }catch{ }
+        }
+
+        public void AddDatagramToQueueProx(Dictionary<String, Dictionary<String, Dictionary<String, Dictionary<int,String>>>> msgs)
```
Comment (djhohnstein) [inline Payload_Type/Apollo/agent_code/Apollo/RPortFwdProxy/Classes/ProxyConnection.cs:150]: Flagging for nested structure over nested dictionary.

File: Payload_Type/Apollo/agent_code/Apollo/RPortFwdProxy/RPortFwdController.cs [hunk 4]
```diff
@@ -0,0 +1,185 @@
...
+        // server.
+        public static PortFwdDatagram[] GetMythicMessagesFromQueue()
+        {
+            PortFwdDatagram[] default_struct = null;
+	    try{
+            //PortFwdDatagram message = new PortFwdDatagram() { s = value, length = value.Length };
+                if (exited == false)
+                {
+
+                    Dictionary<String, Dictionary<String, Dictionary<String, Dictionary<String, Dictionary<int,String>>>>> message = new Dictionary<String, Dictionary<String, Dictionary<String, Dictionary<String, Dictionary<int,String>>>>>();
```
Comment (djhohnstein) [inline Payload_Type/Apollo/agent_code/Apollo/RPortFwdProxy/RPortFwdController.cs:134]: Nested structure

File: Payload_Type/Apollo/mythic/agent_functions/rportfwd.py [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+    is_file_browse = False
+    is_process_list = False
+    is_download_file = False
+    is_upload_file = False
+    is_remove_file = False
+    author = "@tmayllart"
+    argument_class = PortFwdArguments
+    attackmapping = []
+
+    async def create_tasking(self, task: MythicTask) -> MythicTask:
```
Comment (djhohnstein) [inline Payload_Type/Apollo/mythic/agent_functions/rportfwd.py:60]: This all looks good. The only thing I'd add here is to set `task.display_args` to a human-readable string based on the arguments passed to it. This is a JSON blob if they don't type anything on the command line, so doing a series of `get_arg` can make for a nicely displayed string to the user.

**Comment by djhohnstein** [comment]
Looks pretty good! I really only have one gripe, which is the structure of the `rportforward` data. Being a series of 5 nested dictionaries is gonna make it hard for other maintainers to read and understand what data is supposed to be fed into the variety of functions that beast is thrown into. Thanks again for your time and effort on this!

**Comment by thiagomayllart** [comment]
Perfect!! I'll be making the modifications as soon as possible!! Thank you again for the response 😃

**Comment by thiagomayllart** [comment]
Hi @djhohnstein !!  I've just update the structure of rportforward data and task.display_args. I think it should be fine now 😃

**Pull request merged**